### PR TITLE
Rotate TLS ticket "keys"

### PR DESCRIPTION
**Pull request by wmark**

### CODE REVIEW

**Comment by wmark** [comment]
Not sure about the coding style.

This needs pedantic review.

**Comment by mholt** [comment]
I'll give this a good, thorough look -- I invite others who are interested/skilled in this area to do the same! Thanks Mark!

**Comment by wmark** [comment]
I don't know if I've written about that somewhere, but you can do the entire ticket key management without any locks. (There are some on Golang's side.) The tickets need to fit within cache lines (64B) and/or in large registers (amd64: for SSE4, AVXn…). We cannot accomplish this that way, though, because vanilla Golang does not export the actual keys.

(And, in vanilla Golang {ticketName, aesKey, hmacKey} are siblings under the output of SHA512, which begs to be improved.)

**Comment by mholt** [comment]
I realize the server is devoid of tests at the moment, but would a test be feasible to add for this change?

**Comment by wmark** [comment]
We would need to implement mockups of tls.Config and time.Ticker, create the latter outside of the goroutine. I could do that, but it will grow LOC for little benefit. Your call.

**Comment by mholt** [comment]
Since a lot of the server package will be changed around soon and will (hopefully?) be easier to test after that, I can let this go through without a test -- but just so I know, what have you done to ensure this is working? (I did find [this guide](https://jacob.hoffman-andrews.com/README/how-to-check-for-tls-ticket-key-rotation/) which seems instructive.)

Also, it's really interesting that Apache and nginx don't enable this by default. [Apparently even when it is enabled, it's done wrong](https://www.reddit.com/r/netsec/comments/2mkupe/the_sad_state_of_serverside_tls_session/cm58ph1) because of the multi-process model; so implementing this directly into Caddy like this is a huge win, I think. Thank you!

**Comment by wmark** [comment]
I've copied the function's core to Go playground and played with it. It's really just a slice. :-)

Everything else related to tickets is responsibility of Go's tls.*, and that's not many lines to read.

File: server/server.go [hunk 1]
```diff
@@ -378,6 +395,62 @@ func setupClientAuth(tlsConfigs []TLSConfig, config *tls.Config) error {
...
+//
+// Lack of entropy for the very first ticket key results in the feature being disabled (as does Go),
+// later lack of entropy temporarily disables ticket key rotation.
+// Old ticket keys are still phased out, though.
+func standaloneTLSTicketKeyRotation(c *tls.Config, exitChan chan struct{}) {
+	timer := time.NewTicker(tlsNewTicketEvery)
+	defer timer.Stop()
+	// The entire page should be marked as sticky, but Go cannot do that
+	// without resorting to syscall#Mlock. And, we don't have madvise (for NODUMP), too. ☹
+	keys := make([][32]byte, 1, tlsNumTickets)
```
Comment (mholt) [inline server/server.go]: Perhaps these platform-dependent system calls can be a TODO item. Indeed, it'd be nice to protect the keys against memory dumps (etc).

Comment (wmark) [inline server/server.go]: You're right! We can accomplish that later by creating a new file with ```+build linux,!appengine``` or the like.

But that's only half the story: Go eventually copies those values, and we cannot protect them there.

File: server/server.go [hunk 2]
```diff
@@ -378,6 +395,62 @@ func setupClientAuth(tlsConfigs []TLSConfig, config *tls.Config) error {
...
+
+	rng := c.Rand
+	if rng == nil {
+		rng = rand.Reader
+	}
+	if _, err := io.ReadFull(rng, keys[0][:]); err != nil {
+		c.SessionTicketsDisabled = true // bail if we don't have the entropy for the first one
```
Comment (mholt) [inline server/server.go]: Is it safe to set this in a separate goroutine? (Any output from the race detector?)

File: server/server.go (hunk 1 above)
Comment (mholt) [inline server/server.go]: I guess in the meantime we can just encourage the security-conscious to disable swap if possible. (Who wants their web server to be swapping anyway?)

Comment (mholt) [inline server/server.go]: Gah, GitHub doesn't update the page with new comments, even while I'm still reading the code.

Anyway, cool. We can do that later.

File: server/server.go (hunk 2 above)
Comment (wmark) [inline server/server.go]: L419 is racy. (No output.) The worst-case side-effects are harmless, though: Go fails at that too, and disables the feature before us, or a random key is set, but the feature disabled by L419 – the RNG would be unreliable anyway.

**Comment by mholt** [comment]
If you could actually test it out using a tool to help verify the ticket rotation is working (as in the article I linked to above), that would be superb. :+1: And of course, you can set the rotation time down to a much lower value (like a minute?) to do your testing.

File: server/server.go (hunk 2 above)
Comment (mholt) [inline server/server.go]: Good point; maybe make that a comment in the code. That way, if the race detector raises the alarm in the future, the programmer will understand this.

Comment (wmark) [inline server/server.go]: I'd rather like someone who uses are race detector stumble upon this and review it again.

**Comment by mholt** [comment]
Excellent! :raised_hands: Cheers, thank you again.

**Pull request merged**